File: component-acgschedule/src/main/java/com/rabtman/acgschedule/mvp/presenter/ScheduleMainPresenter.java
package com.rabtman.acgschedule.mvp.presenter;

import android.Manifest.permission;
import android.text.TextUtils;
import com.rabtman.acgschedule.R;
import com.rabtman.acgschedule.mvp.contract.ScheduleMainContract;
import com.rabtman.acgschedule.mvp.model.jsoup.DilidiliInfo;
import com.rabtman.common.base.CommonSubscriber;
import com.rabtman.common.base.mvp.BasePresenter;
import com.rabtman.common.di.scope.FragmentScope;
import com.rabtman.common.utils.RxUtil;
import com.tbruyelle.rxpermissions2.RxPermissions;
import io.reactivex.annotations.NonNull;
import io.reactivex.functions.Consumer;
import javax.inject.Inject;

/**
 * @author Rabtman
 */

@FragmentScope
public class ScheduleMainPresenter extends
    BasePresenter<ScheduleMainContract.Model, ScheduleMainContract.View> {

  @Inject
  public ScheduleMainPresenter(ScheduleMainContract.Model model,
      ScheduleMainContract.View rootView) {
    super(model, rootView);
  }

  public void getDilidiliInfo() {
    addSubscribe(
        mModel.getDilidiliInfo()
            .compose(RxUtil.<DilidiliInfo>rxSchedulerHelper())
            .subscribeWith(new CommonSubscriber<DilidiliInfo>(mView) {
              @Override
              public void onComplete() {
                mView.hideLoading();
              }

              @Override
              public void onError(Throwable e) {
                super.onError(e);
                mView.showPageError();
              }

              @Override
              public void onNext(DilidiliInfo dilidiliInfo) {
                mView.showDilidiliInfo(dilidiliInfo);
                mView.showPageContent();
              }
            })
    );
  }

  /**
   * 视频观看权限申请
   */
  public void checkPermission2ScheduleVideo(RxPermissions rxPermissions, final String videoUrl) {
    if (TextUtils.isEmpty(videoUrl)) {
      mView.showError(R.string.msg_error_url_null);
      return;
    }
    rxPermissions.request(permission.WRITE_EXTERNAL_STORAGE,
        permission.READ_PHONE_STATE,
        permission.ACCESS_NETWORK_STATE,
        permission.ACCESS_WIFI_STATE)
        .subscribe(new Consumer<Boolean>() {
          @Override
          public void accept(@NonNull Boolean aBoolean) throws Exception {
            if (aBoolean) {
              mView.start2ScheduleVideo(videoUrl);
            } else {
              mView.showError(R.string.msg_error_check_permission);
            }
          }
        });
  }
}

File: component-acgschedule/src/main/java/com/rabtman/acgschedule/mvp/presenter/ScheduleMainPresenter.kt
package com.rabtman.acgschedule.mvp.presenter

import android.Manifest.permission
import com.rabtman.acgschedule.R
import com.rabtman.acgschedule.mvp.contract.ScheduleMainContract
import com.rabtman.acgschedule.mvp.model.jsoup.ScheduleInfo
import com.rabtman.business.base.CommonSubscriber
import com.rabtman.common.base.mvp.BasePresenter
import com.rabtman.common.di.scope.FragmentScope
import com.rabtman.common.utils.RxUtil
import com.tbruyelle.rxpermissions2.RxPermissions
import javax.inject.Inject

/**
 * @author Rabtman
 */
@FragmentScope
class ScheduleMainPresenter @Inject constructor(model: ScheduleMainContract.Model,
                                                rootView: ScheduleMainContract.View) : BasePresenter<ScheduleMainContract.Model, ScheduleMainContract.View>(model, rootView) {
    fun getScheduleInfo() {
        addSubscribe(
                mModel.getScheduleInfo()
                        .compose(RxUtil.rxSchedulerHelper())
                        .subscribeWith(object : CommonSubscriber<ScheduleInfo>(mView) {
                            override fun onComplete() {
                                mView.hideLoading()
                            }

                            override fun onError(e: Throwable) {
                                super.onError(e)
                                mView.showPageError()
                            }

                            override fun onNext(scheduleInfo: ScheduleInfo) {
                                mView.showScheduleInfo(scheduleInfo)
                                mView.showPageContent()
                            }
                        })
        )
    }

    /**
     * 视频观看权限申请
     */
    fun checkPermission2ScheduleVideo(rxPermissions: RxPermissions, videoUrl: String?) {
        videoUrl?.takeIf { it.isNotBlank() }?.let {
            rxPermissions.request(permission.WRITE_EXTERNAL_STORAGE,
                    permission.READ_PHONE_STATE,
                    permission.ACCESS_NETWORK_STATE,
                    permission.ACCESS_WIFI_STATE)
                    .subscribe { aBoolean ->
                        if (aBoolean) {
                            mView.start2ScheduleVideo(it)
                        } else {
                            mView.showError(R.string.msg_error_check_permission)
                        }
                    }
        } ?: run {
            mView.showError(R.string.msg_error_url_null)
        }
    }
}